### PR TITLE
ROOT-8575: Generate dictionary sources after creating its library target and inherit target properties using generator expressions

### PR DESCRIPTION
**Pull request by amadio**
This will allow the dictionary target to inherit properties such as include directories, etc, from its associated library target.

### CODE REVIEW

**Comment by amadio** [comment]
@phsft-bot build

**Comment by amadio** [comment]
@phsft-bot build

File: cmake/modules/RootMacros.cmake [hunk 1]
```diff
@@ -484,8 +484,17 @@ function(ROOT_GENERATE_DICTIONARY dictionary)
...
   get_filename_component(dictname ${dictionary} NAME)
 
-  #---roottest compability
-  add_custom_target(${dictname} DEPENDS ${dictionary}.cxx ${pcm_name} ${rootmap_name} ${cpp_module_file})
+  if(ARG_NOTARGET)
+    if(TARGET ${ARG_MODULE})
+      target_sources(${ARG_MODULE} PRIVATE ${dictionary}.cxx)
+    else()
+      message(FATAL_ERROR
+        " When used with NOTARGET, the MODULE option must passed a name of an existing target.\n"
```
Comment (vgvassilev) [inline cmake/modules/RootMacros.cmake]:
```suggestion
        " When used with NOTARGET, the MODULE option must be passed a name of an existing target.\n"
```

**Comment by vgvassilev** [comment]
Would this model work when we are in multidict mode? ROOT has this mode for GenVector and Smatrix.

File: cmake/modules/RootMacros.cmake (hunk 1 above)
Comment (amadio) [reply cmake/modules/RootMacros.cmake]: Thanks, I also noticed it and will submit a new round of changes with this fixed.

**Comment by amadio** [comment]
> Would this model work when we are in multidict mode? ROOT has this mode for GenVector and Smatrix.

Yes, it does work with multidict.

**Comment by amadio** [comment]
This is still not the final version, but should be ready for review now. Only the second to last commit is not in its final form.

**Comment by vgvassilev** [comment]
Great! We should probably squash the N calls (when in multidict mode) into one single call. I am not sure if that should be part of this PR but I think it makes things clearer.

**Comment by amadio** [comment]
Making `ROOT_STANDARD_LIBRARY_PACKAGE()` support multidict will be a different PR, but I am thinking about it. For example, all split libraries are not added as deps in `ROOT_LIBRARY_TARGETS` because that's only changed in `ROOT_STANDARD_LIBRARY_PACKAGE()`, but should probably be changed in `ROOT_LINKER_LIBRARY()` instead.

**Comment by amadio** [comment]
Alright, looking in Jenkins it seems this will now pass all tests. I will reorganize the commits again and push tonight or tomorrow.

**Comment by amadio** [comment]
@phsft-bot build

File: cmake/modules/RootMacros.cmake [hunk 2]
```diff
@@ -479,14 +479,23 @@ function(ROOT_GENERATE_DICTIONARY dictionary)
...
+      set(module_defs $<TARGET_PROPERTY:${ARG_MODULE},COMPILE_DEFINITIONS>)
+    endif()
   endif()
 
   #---call rootcint------------------------------------------
   add_custom_command(OUTPUT ${dictionary}.cxx ${pcm_name} ${rootmap_name} ${cpp_module_file}
                      COMMAND ${command} -v2 -f  ${dictionary}.cxx ${newargs} ${excludepathsargs} ${rootmapargs}
-                                        ${definitions} ${includedirs} ${ARG_OPTIONS} ${headerfiles} ${_linkdef}
+                                        ${definitions} "$<$<BOOL:${module_defs}>:-D$<JOIN:${module_defs},;-D>>"
+                                        ${includedirs} "$<$<BOOL:${module_incs}>:-I$<JOIN:${module_incs},;-I>>"
```
Comment (amadio) [inline cmake/modules/RootMacros.cmake]: @vgvassilev If I put `${includedirs}` after `"$<$<BOOL:${module_incs}>:-I$<JOIN:${module_incs},;-I>>"`, the cxxmodule s build fails, could you help me figure out why? The error is the "redefinition of RCompressionSettings" shown in the last Jenkins failure.

**Pull request merged**